### ISSUE TITLE
Turning on code generation for resource files with cultures in their names results in an error

### ISSUE THREAD

**Issue by davkean** [dotnet/project-system#1553]
1. File -> New -> Console Application (.NET Core)
2. Add New Resource called "Resource.en-US.resx

In the Error List:

```
Error		Custom tool PublicResXFileCodeGenerator failed to produce an output for input file 'Resource.en-US.resx' but did not log a specific error.	ConsoleApp181	c:\users\davkean\documents\visual studio 2017\Projects\ConsoleApp181\ConsoleApp181\Resource.en-US.resx	1	
```

Did some investigation - this appears to be (somewhat) by design by the code generator (seeResXFileCodeGenerator.GenerateCode) and has been it's behaviour since VS 2005 (yes 2005, not 2015). What is different in CPS though, is that it errors in this situation - whereas, the legacy project system outputs an empty source file.

Looking through the bug that caused them to implement this behavior was to avoid having both the neutral and the language-specific resource generate the same code and result in build errors. However, in some situations you don't have a neutral Resource file such as when your neutral resources end up in a satellite assembly. ie I don't think this behavior makes sense.

**Comment by davkean** [comment]
@nguerrera This was cause of your issue when you tried to do this.

**Comment by davkean** [comment]
Okay, thinking about this a little more I understand why that did this - but I don't agree with the fix they made. Basically, they wanted to prevent the situation where you add multiple language-specific resources of the same resource file via Add Item, from getting code generated, as that will result in build errors.

Instead, I think we should fix this in a couple of ways:

1. If it is a language-specific resource and the language-neutral one already exists, we shouldn't turn on code generation for it.
2. We should let you specifically turn on code generation for it if you so want.

**Comment by nbobogk** [comment]
I also have this problem and unfortunately can not find a working solution for this! When I create a new project (.Net Core 2.0 Web application, Visual Studio 2017 Community, latest version) and follow the instructions of the site "https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization" (create resources directory and there for example a file "Resources / Models.ResTestModel.de.resx"), I always get the error message:
"Custom tool ResXFileCodeGenerator failed to make an output for input file ... but did not log a specific error". I have tested it on different systems. This is obviously the case when the language identifier (".de") appears in the file name of the resource file. Files without language recognition seem to work!
Can anyone offer a solution? I would be very grateful for this!

**Comment by bill-call** [comment]
For the sake of clarity, are we to understand that the work-around is to have a language-neutral .resx file in the same directory?  If I understand @davkean correctly, this would cause the static class to be emitted, but would not prevent the error message for each .resx file that had a locale embedded in the file name, which is okay, because the error itself is harmless, to the extent that it does not mean that your build is bad.

**Comment by bill-call** [comment]
Well, I tried the above.  It does seem to stop the errors, and XXX.designer.cs does get generated for the neutral .resx file, but if you look at the generated class, the value of each ID are the same as the ID, even if they are different in the neutral .resx file. Is that right?

Regardless, I can't seem to retrieve the localized text in my view.  I always get the ID back, as if nothing were defined.  This is true even if I manually edit the designer class so that ID and value are distinct.

**Comment by bill-call** [comment]
Figured out why view localization wasn't working; borked the path names.

**Comment by bill-call** [comment]
Okay, for everybody playing along at home, let me see if I can summarize this:

If you create, modify or select (for the first time), or manually run ResXFileCodeGenerator on a .resx file with culture designator in the file name (it has the form like foo.en.resx or foo.en-US.resx) but you do not have a culture-neutral  .resx file of the same base name (foo.resx), then you will get this error.  The error is correct, in that no class file was emitted, but it is wrong, in that it should not be an error. The fact that no class file was emitted is by design; the fact this specific condition is now being flagged as an error is incorrect.

Absent a fix, we have two choices:

1. - We can create a culture-neutral .resx file, which will allow the class to be emitted and thus prevent the error.

2. - We can just ignore the error, as it has no effect on the viability of the build.

**Comment by mne1991** [comment]
Any news regarding this bug? Although it doesn't have an effect on building, it is still not ideal to always have 1 error in error list.

**Comment by Legends** [comment]
Ooooh my, even resources files are rendered useless, don't even want to think about digging deeper into ASP.NET Core 2.0.... Back to MVC 5... until ASP.NET Core v6 perhaps, see you then...

**Comment by davkean** [comment]
@mne1991 Yes, closing the solution and reopen after removing the "Custom Tool" for the language-specific ResX (neutral can still have Custom Tool) in the properties windows. Basically, the legacy project system ignores and avoids showing this error, the new project system shows it. We'll fix it in a future update.

**Comment by RodrigoFCampos** [comment]
I am using windows 10 / iis / .net core 2.1 / latest vs 2017 and I must say that I have 2 resources files: 

```
Resource.resx
Resource.en-US.resx
```

And sometimes (literally sometimes, because I can't reproduce this error when I want) I get the following error after a build:

> Error		Custom tool PublicResXFileCodeGenerator failed to produce an output for input file 'Resources\Resource.en-US.resx' but did not log a specific error.

I personally like the last part of the error "... but did not log a specific error." :)
I must say that this is a really strange error, because when I click on the error (in the "Error List" menu), the Resource file "Resource.en-US.resx" opens, then I close it, and I run another build and this time no errors appear. 

To sum up, well I have no clue of what is happening :P 
Am I missing something?

**Comment by davkean** [comment]
Remove the Custom Tools "PublicResXFileCodeGenerator" from the Resource.en-US.resx, and reload the project. You will not see the error again.

**Comment by tolgabalci** [comment]
Resolved the issue by changing PublicResXFileCodeGenerator to ResXFileCodeGenerator on all language specific resource files.  I left the Resource.resx using the PublicResXFileCodeGenerator because that's how other projects consume my Resource library.

**Comment by brunk01** [comment]
This helped me if anyone is looking for a quick and dirty fis.
https://marketplace.visualstudio.com/items?itemName=DonAirey.ResourceGenerator

**Comment by JohnyWS** [comment]
@tolgabalci but won't you still get an error when clicking the "Run Custom Tool" option for the files? We removed the designer files completely from the culture specific resources, plus removed the custom tool on the resx, so only the neutral language resx files had the tooling to create designer files.

**Comment by macux** [comment]
Would be really good if this error could be removed, it just sent me on a wild goose chase whilst trying to localize data annotations in a dotnet core app (which itself is confusing enough).

**Comment by RoryMcCrossan** [comment]
Still having this issue in VS2019. The RESX editor as a whole is dreadful and badly needs updating.

**Comment by Latency** [comment]
I am having the same issue with VS2019 v16.7.3
Problem still isn't resolved.

Is .resx deprecated for SDK format now?   What is the replacement for this?   I agree with Rory... this should be compatable.

**Comment by davkean** [comment]
@Latency ResX isn't deprecated. Did the [workaround](https://github.com/dotnet/project-system/issues/1553#issuecomment-397600474) not work for you?

**Comment by kiddailey** [comment]
Just wanted to add a note that when this issue occurs in a Xamarin project (VS2019/Win), it appears to disrupt the ability to deploy to a device.  The fix for me in these instances is that I have to disconnect from the paired mac and reconnect.   It doesn't seem to stop deployment to the simulator though.  [The workaround ](https://github.com/dotnet/project-system/issues/1553#issuecomment-397600474) does resolve it for me.

For what it's worth, I always have a culture-neutral resource file in my projects (it's typically the first one I add) and I still experience the issue, so it's not limited to projects that don't already have one.

**Comment by danieldefreitasleite** [comment]
Just set the Access Modifer to No code generation (in fact only the neutral/main resource file will genate the code needed) as showed here: [https://dev.to/krusty93/xamarin-android-and-ios-resx-localization-with-mvvmcross-n02](url).

This way the error should be gone...

**Comment by kYasinAblay** [comment]
i still doesnt work my project :(

**Comment by Didier-Solheid** [comment]
Same for me. A nightmare ...

**Comment by AraHaan** [comment]
I have have just got bitten by this when I have the csproj file itself have the following contents:

```xml
<Project Sdk="Microsoft.NET.Sdk" />
```

While ``Directory.Build.props`` contains all of the ``PropertyGroup`` Items that should be set for the project, ``Directory.Build.targets`` for the ``ItemGroup`` entries, and ``Directory.Packages.props`` for all ``ProjectReference``s and ``PackageReference``s.

Placing the ``EmbeddedResource`` element on an neutral resx file and having it set to ``ResXFileCodeGenerator`` on the generator ends up with the resources not getting generated EVER...

I expected it to work even when placed in ``Directory.Build.targets``.

**Comment by maxkatz6** [comment]
It's 2021, shouldn't ResXFileCodeGenerator be replaced with new source generator?

**Comment by AraHaan** [comment]
I found the source generator myself and you can find how I added a reference to it manually in https://github.com/Elskom/Sdk/

However I would love for the Microsoft.NET.Sdk to include that generator by default so then whenever one tries to make a new resx file it won't go about using ``ResXFileCodeGenerator`` (and possibly remove ``ResXFileCodeGenerator`` at the same time too). But I think that is part of the templates that would need to be updated then as well too.

**Comment by guiggsrm** [comment]
To solve this problem you have to modify the csproj file. An inconsistence when you copy creates the dependency, but a line to refer the design file is missing.
`  <ItemGroup>
    <Compile Update="Properties\Resources.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
    </Compile>
    <Compile Update="Properties\Resources.en-US.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
    </Compile>
  </ItemGroup>`

Look the dependentUpo tag. You have to create a line equals de design file and close the VS. After that everithing works fine;
`  <ItemGroup>
    <Compile Update="Properties\Resources.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
      <DependentUpon>Resources.resx</DependentUpon>
    </Compile>
    <Compile Update="Properties\Resources.en-US.Designer.cs">
      <DesignTime>True</DesignTime>
      <AutoGen>True</AutoGen>
	  <DependentUpon>Resources.en-US.resx</DependentUpon>
    </Compile>
  </ItemGroup>`

**Comment by TheCollegedude** [comment]
> To solve this problem you have to modify the csproj file. An inconsistence when you copy creates the dependency, but a line to refer the design file is missing.
No, does not work for me.

**Comment by ocallesp** [comment]
@ocallesp to check if it still reproes

**Comment by ocallesp** [comment]
This doesn't repro is VS 2022 Main

**Comment by AraHaan** [comment]
I think the proper way to do this (looking at the runtime repo) is to rig up arcade’s xliff tasks and basically have a single universal resx files but have the language specific copies be *.xlf files.

**Comment by Nix1983** [comment]
I get the same error. The problem is the "/" in the value for example "de/counter". How can i made "/" as value?

**Comment by augong0301** [comment]
same issue after my project migrates  from .net framework 48 to .net 8 
the custom .resx never generate from that on.

**Comment by drewnoakes** [comment]
I can no longer reproduce this issue. Adding a "Resources File" with name `Resource.en-US.resx` does not add anything to the `.csproj` file, and the default behaviour in the SDK prevents any code generation from happening.

Adding `Resources.resx` instead does add to the project, as expected.

Closing this as fixed.

**Comment by JosiahDanger** [comment]
I am currently experiencing this issue in Visual Studio 2026 (18.4.3). I think it has resurfaced.

A '*.Designer.cs' file is generated for my 'Text.resx', but not for my 'Text.en-GB.resx'. Let me know if you need more information.

**Comment by drewnoakes** [comment]
@JosiahDanger that's by design. You only have one `.Designer.cs` file for each group of languages. The assumption is that you have a `Text..resx` file alongside `Text.en-GB.resx`. If not, you can set the `Generator` metadata manually on your project item to make it generate. Add a new .resx file to your project and see what XML is generated for that, then update it for your cultured file.

**Comment by JosiahDanger** [comment]
> [@JosiahDanger](https://github.com/JosiahDanger) that's by design. You only have one `.Designer.cs` file for each group of languages. The assumption is that you have a `Text..resx` file alongside `Text.en-GB.resx`. If not, you can set the `Generator` metadata manually on your project item to make it generate. Add a new .resx file to your project and see what XML is generated for that, then update it for your cultured file.

Thank you for your response. That does make sense.

However, I can create a scenario for testing purposes, in which my only '*.resx' file is 'Text.en-GB.resx'. In this scenario, the generic 'Text.resx' does not exist. Under these conditions, the error described by @davkean occurs; Visual Studio fails to generate the expected 'Text.en-GB.Designer.cs' file. Please can you confirm whether or not this is intended behaviour?

**Comment by drewnoakes** [comment]
It comes down to what's in your project file. If you have the right `<Generator>` on your en-GB file, I would expect it to create a source file for you.

**Comment by JosiahDanger** [comment]
> It comes down to what's in your project file. If you have the right `<Generator>` on your en-GB file, I would expect it to create a source file for you.

I will provide you with an example.

This is [my desktop application](https://github.com/JosiahDanger/DateToday/tree/DT-Version-2.0), which is currently in an early stage of development.

Currently, as you may see in my source code, I have a neutral 'Strings.resx', for which 'Strings.Designer.cs' has been generated successfully.

<img width="582" height="677" alt="Image" src="https://github.com/user-attachments/assets/af0a3359-1504-4528-8286-4af98038b77f" />

And here is the relevant Generator in my '.csproj' file.

<img width="810" height="183" alt="Image" src="https://github.com/user-attachments/assets/ecc5b599-9e3a-4353-8052-c002ec30d1ac" />

Let's perform an experiment. I will delete 'Strings.Designer.cs'. Then, I will rename my '.resx' file to 'Strings.en-GB.resx'.

<img width="595" height="686" alt="Image" src="https://github.com/user-attachments/assets/ce884471-da8e-4098-a6df-8fe5b0da5d2f" />

I confirm that this change is reflected in my '.csproj' file.

<img width="885" height="173" alt="Image" src="https://github.com/user-attachments/assets/b4de1025-4e64-4bbb-bbc1-9994bdd620a1" />

Next, I perform a clean and rebuild.

<img width="1254" height="220" alt="Image" src="https://github.com/user-attachments/assets/bbe8bdd1-4d3b-490c-b08e-f2e170ad499c" />

The error occurs. I believe this is a bug.

**Comment by drewnoakes** [comment]
I see, thanks for clarifying. Then this is a bug in the generator itself rather than the project system. There are a few different copies of that generator across various parts in the IDE unfortunately.

I would suggest trying the newer ResX source generator. The legacy approach requires a tool like VS to generate the designer file for you, and then you have to check the file in. The source generator works everywhere the compiler works, generates better code, and you don't check that code in.

**Comment by JosiahDanger** [comment]
> I see, thanks for clarifying. Then this is a bug in the generator itself rather than the project system. There are a few different copies of that generator across various parts in the IDE unfortunately.
> 
> I would suggest trying the newer ResX source generator. The legacy approach requires a tool like VS to generate the designer file for you, and then you have to check the file in. The source generator works everywhere the compiler works, generates better code, and you don't check that code in.

Thank you for your response. To whom may I report this bug? And how may I use the newer ResX source generator?

**Comment by drewnoakes** [comment]
It's in package `Microsoft.CodeAnalysis.ResxSourceGenerator`:

https://www.nuget.org/packages/Microsoft.CodeAnalysis.ResxSourceGenerator/5.0.0-1.25277.114

You could file a feedback ticket on developer community, but I wouldn't expect it to get much traction. The source generator is a better option for you in several ways.

**Comment by AraHaan** [comment]
> It's in package `Microsoft.CodeAnalysis.ResxSourceGenerator`:
> 
> https://www.nuget.org/packages/Microsoft.CodeAnalysis.ResxSourceGenerator/5.0.0-1.25277.114
> 
> You could file a feedback ticket on developer community, but I wouldn't expect it to get much traction. The source generator is a better option for you in several ways.

While that is true, I wish it also handled ``Image``'s well as also for Windows Forms/WPF applications as well that target ``netX.Y-windows``...

Currently I must work around it by including svg's in the resx file and then use an svg library to convert the svg to ``Image`` (System.Drawing.Image)'s in order to do what I need to do...
